File: src/main/scala/kim/dencka/scala/easy/ReverseLinkedList.scala
package kim.dencka.scala.easy

import kim.dencka.scala.payloads.LinkedListStructure._

object ReverseLinkedList extends App {

  def reverseList(head: ListNode): ListNode = {
    @annotation.tailrec
    def run(ln: ListNode, acc: ListNode): ListNode = ln match {
      case null => acc
      case _ => run(ln.next, new ListNode(ln.x, acc))
    }
    run(head, null)
  }

}

File: src/main/scala/kim/dencka/java/easy/ReverseLinkedList.java
package kim.dencka.java.easy;

import kim.dencka.java.payloads.ListNode;

class ReverseLinkedList {
    public static void main(String[] args) {
        ListNode input = ListNode.fromArray(1, 2, 3, 4, 5);
        System.out.println(reverseList(input)); // [5,4,3,2,1]
    }

    // Iterative
    public static ListNode reverseList(ListNode head) {
        ListNode prev = null;
        while (head != null) {
            ListNode next = head.next;
            head.next = prev;
            prev = head;
            head = next;
        }
        return prev;
    }

    // Recursive
    public static ListNode reverse(ListNode list, ListNode acc) {
        if (list == null) return acc;
        else return reverse(list.next, new ListNode(list.val, acc));
    }

}
